### ISSUE TITLE
Bug: Scale is defined as i32

### ISSUE THREAD

**Issue by WhyNotHugo** [Smithay/client-toolkit#171]
Throughout the codebase, the `scale` is defined as `i32`.  
This makes it impossible to properly obtain or set a scale, since compositors don't use integers for this. As a result, applications using `client-toolkit` are rendered blurry.

For example, on my current setup:

```console
$ swaymsg -t get_outputs
Output eDP-1 'Sharp Corporation 0x148B 0x00000000'
  Current mode: 3840x2160 @ 59.997002 Hz
  Scale factor: 2.500000...

Output DP-2 'Ancor Communications Inc ROG PG348Q #ASMuhJCcVsHd' (focused)
  Current mode: 3440x1440 @ 59.973000 Hz
  Scale factor: 1.500000
...
```

I took a quick look at sway's codebase, and it seems to use `double` or `float` for scaling factors (although defined simply at `1.5` in the config).

**Comment by vberger** [comment]
The integer scaling factors are [a constraint of the wayland protocol itself](https://gitlab.freedesktop.org/wayland/wayland/-/blob/master/protocol/wayland.xml#L2612), which only support integer scaling factors. All compositors that expose fractional scaling do so by rescaling themselves the content drawn by the clients.

See for example the sway config manual page:

> `output <name> scale <factor>`
> Scales the specified output by the specified scale factor. An integer is recommended, but fractional values are also supported. **If a fractional value are specified, be warned that it is not possible to faithfully represent the contents of your windows - they will be rendered at the next highest integer scale factor and downscaled.** You may be better served by setting an integer scale factor and adjusting the font size of your applications to taste. HiDPI isn't supported with Xwayland clients (windows will blur).

The question of supporting fractional scaling at the protocol level is debate between the Wayland developers and the compositor developers that has literally been going on for years, and SCTK cannot really do anything about that.

**Comment by WhyNotHugo** [comment]
Hmm, interesting. I was unaware of the limitation.

Seems to be a silly one: without fractional scaling I would be unable to use my current hardware with Wayland and would need to replace all my screens with very specific ones (likely very hard to find on the market). I use a scaling of `1.5`. `1` is too low and results in text I can't read (bad eyesight). `2` is just too little content on screen.

Back on topic though: how is it that other applications deal with this? E.g.: SCTK seems to pick up a lower scaling factor and render blurry. Or is maybe the issue elsewhere?

**Comment by vberger** [comment]
The issue is likely somewhere else: SCTK does not render the content of your windows. All it does is pick the (integer) scaling factor advertised by the server (which should be the rounded up version of the fractional one you configured for each screen), and forwards it to the app.

The app is then responsible to draw its contents in an HiDPI-aware way, by increasing the size of its drawing buffer to the correct size, and telling the compositor it did so.

So, for example, if your app is supposed to be drawn as a 200x200 logical pixels app (which is the size given to you as [`Event::Configure`](https://docs.rs/smithay-client-toolkit/0.12.1/smithay_client_toolkit/window/enum.Event.html) by the `Window` helper. At the moment of actually drawing you should:

1. check what is the scaling factor appropriate for the `WlSurface` you are going to draw on using [`sctk::get_surface_scale_factor`](https://docs.rs/smithay-client-toolkit/0.12.1/smithay_client_toolkit/fn.get_surface_scale_factor.html). Let's assume you get 3.
2. Create a buffer of 600x600 pixels (3 * 200), and draw your contents on it with a 3x scale factor
3. Notify the server that you are drawing your contents at a 3x scale factor using [`WlSurface::set_buffer_scale`](https://docs.rs/wayland-client/0.28.2/wayland_client/protocol/wl_surface/struct.WlSurface.html#method.set_buffer_scale)
4. Submit your buffer & commit the surface

If you don't notify the server what scale factor you are drawing with, it'll interpret it as you drawing with a scale factor of 1, and thus upscale your contents to the scale factor of the relevant screen, resulting in blurriness.

**Comment by WhyNotHugo** [comment]
Thanks once again for a very thorough and clear explanation!

So what I'm trying to do right now is patch `yofi` to actually respect my compositor's scaling settings. After some inspecting and debugging, it seems that the result of [its call to `sctk::get_surface_scale_factor`](https://github.com/l4l/yofi/blob/79cb8dddf605873c6910b4832755cc35c0e1914b/src/surface.rs#L58-L60) is always `1` (even though the active display is `1.5`. Shouldn't this be `2`?

I've even added this:
```rust
        let reported_scale: u16 = sctk::get_surface_scale_factor(&surface)
            .try_into()
            .expect("invalid surface scale factor");
        println!("scale is {}", reported_scale);
```

Which yields this:

```
scale is 1
```

Am I missing something here?

**Comment by vberger** [comment]
Was the `WlSurface` created using [`Environment::create_surface()`](https://docs.rs/smithay-client-toolkit/0.12.1/smithay_client_toolkit/environment/struct.Environment.html#method.create_surface) ? This is necessary for `set_surface_scale_factor` to work.

**Comment by WhyNotHugo** [comment]
The above used `env.create_surface().detach();` (the whole file is in the above link)

I did try this too, and it also returns `1`:

```rust
        let surface0 = env.create_surface();
        let layer_shell = env.require_global::<zwlr_layer_shell_v1::ZwlrLayerShellV1>();

        let scale2: u16 = sctk::get_surface_scale_factor(&surface0)
            .try_into()
            .expect("invalid surface scale factor");
        println!("scale is {}", scale2);
```

**Comment by vberger** [comment]
Oh, you should check the scale every time you redraw rather than once at the start.

The client app (and thus SCTK) cannot know the correct value for the scale factor until it knows which monitor(s) your surface is being displayed on. And it cannot know that until the surface is actually displayed (due to how the protocol works), which again cannot happen until something has actually been drawn.

So yes, the first frame will always be rendered at a scale factor of 1, [this is a known issue of the protocol](https://gitlab.freedesktop.org/wayland/wayland/-/issues/133).

Note that checking the scale factor using `sctk::get_surface_scale_factor` is very cheap (it just reads a precomputed value), so there is no performance concerns in just checking it before every redraw. If you need to manage your buffers more efficiently you can also use [`Environment::create_surface_with_scale_callback`](https://docs.rs/smithay-client-toolkit/0.12.1/smithay_client_toolkit/environment/struct.Environment.html#method.create_surface_with_scale_callback) to provide a callback that will be invoked whenever the scale factor changes.

**Comment by l4l** [comment]
Is it possible to add additional draw, specifically for getting the proper scale factor then? If I simply damage and commit the surface right before the main loop `EventLoop::dispatch` fails with "Broken pipe".

That in particular can be reproduced in my branch: https://github.com/l4l/yofi/commits/hidpi-update-fix-broken-pipe. Although, it's not really a minimal example but all the wayland-related stuff goes in changed main.rs and [surface.rs](https://github.com/l4l/yofi/blob/f215106138a931302557f50e25d2886215bda1d6/src/surface.rs). May I ask you to look at this?

**Comment by vberger** [comment]
> Is it possible to add additional draw, specifically for getting the proper scale factor then? If I simply damage and commit the surface right before the main loop EventLoop::dispatch fails with "Broken pipe".

I suspect you also get a "protocol error" message on stderr before the panic message, right?

Indeed, currently there is a need to have an initial draw on some real content: you can't get the scale factor if your surface is not displayed, and you surface is not displayed if it has no content.

You can set up some heuristics though, like if all outputs have the same scale factor, then you can guess what your scale factor will be. But otherwise, the issue is that you cannot in general guess which output your surface will be displayed on.

**Comment by l4l** [comment]
> I suspect you also get a "protocol error" message on stderr before the panic message, right?

Not really, there's no other output at stout/stderr/log. Perhaps it should be enabled somehow?

**Comment by WhyNotHugo** [comment]
@vberger Thanks for the very thorough explanation, this has been very educational.

> So yes, the first frame will always be rendered at a scale factor of 1, this is a known issue of the protocol.

Huh, I've also noticed this quirk a few times on iPads; I wonder if they suffer the same limitations at some level.

---

I wonder if you've ever considered abstracting the scaling to a degree and handling that internally.

For example, if I create a 200x200px widget on Qt, and my scaling factor is 3x, Qt will make this a 600x600px widget and handle all the "multiple by 3" internally. So app developers don't even need to think about scaling at all, it's completely transparent.

Do you think an approach like this might make sense? I can image it'd help applications avoid having to get the scale and multiply every single dimension manually.

**Comment by l4l** [comment]
> For example, if I create a 200x200px widget on Qt, and my scaling factor is 3x, Qt will make this a 600x600px widget and handle all the "multiple by 3" internally. So app developers don't even need to think about scaling at all, it's completely transparent.
> 
> Do you think an approach like this might make sense? I can image it'd help applications avoid having to get the scale and multiply every single dimension manually.

I don't really think it may work there. Qt handles all of the drawing by itself as far as I remember and on the other hand sctk is only provides a raw buffer of specified size where you may draw by yourself.

**Comment by vberger** [comment]
@l4l
> Not really, there's no other output at stout/stderr/log. Perhaps it should be enabled somehow?

Okay, *that* is surprising. Do you have a commit I can fetch to try and reproduce that on my computer? Is that the branch you linked earlier?

@WhyNotHugo
> Do you think an approach like this might make sense?

As @l4l mentioned, the fact that SCTK does not handle drawing at all make this more difficult to manage.

The way the Wayland protocol considers HiDPI scaling is that everything is managed in logical coordinates (input events, position of subsurfaces / popups, etc). As a result, for a purely native Wayland app, the only part of the code that needs to be aware of scaling factor is the drawing logic (which is a part that is as of yet not covered at all by SCTK, and would be quite difficult to given the current scope of the library). The rest of the app can always behave as if the scaling factor was not a thing.

I understand that this way of handling things is not the norm among different platforms, and that it may force to do a lot of conversion in cross-platform libraries (winit is an example of that), but SCTK is essentially a toolbox for making wayland apps, so it seems better to me to stick with the Wayland approach here.

However, SCTK's API is by no means complete nor final, if you can think of some changes or improvements to it that could make things easier, then I am all ears.

**Comment by l4l** [comment]
> Do you have a commit I can fetch to try and reproduce that on my computer? Is that the branch you linked earlier?

Yes, that's the same: https://github.com/l4l/yofi/commits/hidpi-update-fix-broken-pipe. A minor addition: stderr/stdout for compositor (sway in my case) is also empty.

**Comment by vberger** [comment]
Okay, got it. It was indeed a protocol error, but apparently the rust backend of wayland-client does not handle it correctly, so that's a bug that I'll need to fix.

So, the protocol error is actually completely unrelated to hidpi handling, but related to the "initial draw":

```
zwlr_layer_surface_v1@17: error 2: layer_surface has never been configured
```

This is because the layer-shell protocol (like the xdg-shell protocol) requires you to first commit the `wl_surface` once with no buffer at all after creating the associated `zwlr_layer_surface`, to signal the compositor that you intent to create a layer surface. Which you did according to the `WAYLAND_DEBUG` log I'm reading.

However, once that is done, you must wait for the compositor to reply with a `zwlr_layer_surface.configure` event and ack the associated `wl_surface.configure` event before drawing your first frame, using the size given to you by the compositor in this event.

**Comment by l4l** [comment]
Does that mean the code shouldn't work even without patch?

**Comment by vberger** [comment]
If I understand it correctly (I only gave a quick look), the code without the last commit should be correct.

You enter your main loop with the surface not yet existing because you have not drawn anything yet. Then you'll receive the configure event from the server, which your code interprets as necessitating a redraw, which then causes you to draw your first frame.

That frame uses a scale factor of 1 though, so as you're doing lazy drawing, you'll probably want to use [`create_surface_with_scale_callback`](https://docs.rs/smithay-client-toolkit/0.12.1/smithay_client_toolkit/environment/struct.Environment.html#method.create_surface_with_scale_callback), so that SCTK will notify you if the scale factor changes (which will be right after your first draw is the monitor has a scale factor greater than 1), and you cans schedule a second redraw, which will be at the correct scale factor.

**Comment by vberger** [comment]
So, I'm going to assume this is done an can be closed. :wink:
